src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs

```diff
@@ -60,536 +61,319 @@ public class StreamSubscription : ISubscriptionEnumerator {
+									_channel.Writer.TryComplete(ex);
+									return;
+								}
 
-						if (message is ClientMessage.NotHandled notHandled &&
-						    RpcExceptions.TryHandleNotHandled(notHandled, out var ex)) {
-							_subscriptionConfirmed.TrySetException(ex);
-							return;
-						}
+								if (!(message is ClientMessage.ReadStreamEventsForwardCompleted completed)) {
+									_channel.Writer.TryComplete(
+										RpcExceptions.UnknownMessage<ClientMessage.ReadStreamEventsForwardCompleted>(
+											message));
+									return;
+								}
 
-						switch (message) {
-							case ClientMessage.SubscriptionConfirmation confirmed:
-								var caughtUp = StreamRevision.FromInt64(Math.Max(confirmed.LastEventNumber.Value + 1,
-									currentStreamRevision.ToInt64()));
-								Log.Verbose(
-									"Live subscription {subscriptionId} to {streamName} confirmed at {streamRevision}.",
-									_subscriptionId, _streamName, caughtUp);
-								_subscriptionConfirmed.TrySetResult(caughtUp);
-								subscriptionStarted.TrySetResult(true);
-								ReadHistoricalEvents(currentStreamRevision);
+								switch (completed.Result) {
+									case ReadStreamResult.Success:
+										if (completed.Events.Length == 0 && completed.IsEndOfStream) {
+											NotifyCaughtUp(StreamRevision.FromInt64(completed.FromEventNumber));
+											return;
+										}
+
+										foreach (var @event in completed.Events) {
+											var streamRevision = StreamRevision.FromInt64(@event.OriginalEventNumber);
+											if (streamRevision > caughtUp) {
+												NotifyCaughtUp(streamRevision);
+												return;
+											}
+
+											await _channel.Writer.WriteAsync(@event, _cancellationToken)
+												.ConfigureAwait(false);
+										}
+
+										ReadHistoricalEvents(StreamRevision.FromInt64(completed.NextEventNumber));
+
+										void NotifyCaughtUp(StreamRevision streamRevision) {
+											Log.Verbose(
+												"Live subscription {subscriptionId} to {streamName} caught up at {streamRevision} because the end of stream was reached.",
+												_subscriptionId, _streamName, streamRevision);
+											caughtUpSource.TrySetResult(caughtUp);
+										}
 
-								return;
-							case ClientMessage.SubscriptionDropped dropped:
-								Log.Debug(
-									"Live subscription {subscriptionId} to {streamName} dropped: {droppedReason}",
-									_subscriptionId, _streamName, dropped.Reason);
-								switch (dropped.Reason) {
-									case SubscriptionDropReason.AccessDenied:
-										Fail(RpcExceptions.AccessDenied());
 										return;
-									case SubscriptionDropReason.NotFound:
-										Fail(RpcExceptions.StreamNotFound(streamName));
+									case ReadStreamResult.NoStream:
+										Log.Verbose(
+											"Live subscription {subscriptionId} to {streamName} stream not found.",
+											_subscriptionId, _streamName);
+										ReadHistoricalEvents(startRevision);
```

Comment:
Should we pause for a few milliseconds before reading again here?

Reply:
Yes